service/src/java/org/apache/hive/service/cli/operation/OperationLogManager.java

```diff
@@ -0,0 +1,387 @@
+        return;
+      }
+    }
+    HISTORIC_OPERATION_LOG_ROOT_DIR = historicLogLoc;
+  }
+
+  public static OperationLog createOperationLog(Operation operation, QueryState queryState) {
+    HiveSession session = operation.getParentSession();
+    File parentFile = session.getOperationLogSessionDir();
+    boolean isHistoricLogEnabled = HISTORIC_OPERATION_LOG_ROOT_DIR != null;
+    if (isHistoricLogEnabled && operation instanceof SQLOperation) {
+      String sessionId = session.getSessionHandle().getHandleIdentifier().toString();
+      parentFile = new File(HISTORIC_OPERATION_LOG_ROOT_DIR + "/" + sessionId);
+      if (!parentFile.exists()) {
+        if (!parentFile.mkdirs()) {
+          LOG.warn("Unable to create the historic operation log session dir: " + parentFile +
+              ", fall back to the original operation log session dir.");
+          parentFile = session.getOperationLogSessionDir();
+          isHistoricLogEnabled = false;
+        }
+      } else if (!parentFile.isDirectory()) {
+        LOG.warn("The historic operation log session dir: " + parentFile + " is exist, but it's not a directory, " +
+            "fall back to the original operation log session dir.");
+        parentFile = session.getOperationLogSessionDir();
+        isHistoricLogEnabled = false;
+      }
+    }
+
+    OperationHandle opHandle = operation.getHandle();
+    File operationLogFile = new File(parentFile, queryState.getQueryId());
+    OperationLog operationLog;
+    HiveConf.setBoolVar(queryState.getConf(),
+        HiveConf.ConfVars.HIVE_SERVER2_HISTORIC_OPERATION_LOG_ENABLED, isHistoricLogEnabled);
+    if (isHistoricLogEnabled) {
+      // dynamically setting the log location to route the operation log
+      HiveConf.setVar(queryState.getConf(),
+          HiveConf.ConfVars.HIVE_SERVER2_LOGGING_OPERATION_LOG_LOCATION, HISTORIC_OPERATION_LOG_ROOT_DIR);
+      if (HiveConf.getBoolVar(queryState.getConf(), HiveConf.ConfVars.HIVE_IN_TEST)) {
+        HiveConf.setBoolVar(queryState.getConf(), HiveConf.ConfVars.HIVE_TESTING_REMOVE_LOGS, false);
+      }
+      LOG.info("The operation log location changes from {} to {}.", new File(session.getOperationLogSessionDir(),
+          queryState.getQueryId()), operationLogFile);
+    }
+    operationLog = new OperationLog(opHandle.toString(), operationLogFile, queryState.getConf());
+    return operationLog;
+  }
+
+  private Set<String> getLiveSessions() {
+    Collection<HiveSession> hiveSessions = sessionManager.getSessions();
+    Set<String> liveSessions = new HashSet<>();
+    for (HiveSession session : hiveSessions) {
+      liveSessions.add(session.getSessionHandle().getHandleIdentifier().toString());
+    }
+    return liveSessions;
+  }
+
+  private Set<String> getHistoricSessions(String logRootDir) {
+    File logDir = new File(logRootDir);
+    Set<String> results = new HashSet<>();
+    if (logDir.exists() && logDir.isDirectory()) {
+      File[] subFiles = logDir.listFiles();
```

Comment:
What happens when we have multiple HS2 instances with the same root dir configuration?

Reply:
Also plan for concurrent file listings and removals

Reply:
oh yes, if multiple instances share the same root dir,  the cleaner will remove the live operation log in other instances. Make the root dir be private by adding a subdir that specific for the instance: the thrift port and the start time, this can avoid concurrent file listings and removals.  Thanks very much for the comments!